consult.el

```diff
@@ -1102,46 +1102,81 @@ for which the command history is used."
+            (symbol-value minibuffer-history-variable))
+          'minibuffer-history-position))
    ;; Otherwise we use a mode-specific history, see `consult-mode-histories'.
-   (t (when-let (history
-                 (or (seq-find (lambda (ring)
-                                 (and (derived-mode-p (car ring))
-                                      (boundp (cdr ring))))
-                               consult-mode-histories)
-                     (user-error
-                      "No history configured for `%s', see `consult-mode-histories'"
-                      major-mode)))
-        (symbol-value (cdr history))))))
+   (t (cl-destructuring-bind (history index-var)
+          (or (cdr (seq-find (lambda (ring)
+                               (and (derived-mode-p (car ring))
+                                    (boundp (cadr ring))))
+                             consult-mode-histories))
+              (user-error
+               "No history configured for `%s', see `consult-mode-histories'"
+               major-mode))
+        (cons (symbol-value history)
+              index-var)))))
+
+(defun consult--label-and-delete-dups (entries)
+  "Label ENTRIES with history indices.
+
+If there are duplicates, the most recent occurrence of the
+duplicate gets labeled with its index, and the others are not
+shown."
+  (let ((ht (make-hash-table :test 'equal))
+        (idx 0)
+        entry
+        accum)
+    (while (setq entry (pop entries))
+      (unless (gethash entry ht)
+        (puthash entry t ht)
+        (push (cons entry (cons idx entry))
+              accum))
+      (cl-incf idx))
+    (nreverse accum)))
 
 (defun consult--history-elements (history)
   "Return elements from HISTORY.
 Can handle lists and rings."
-  (delete-dups (seq-copy (if (ring-p history)
-                             (ring-elements history)
-                           history))))
+  (consult--label-and-delete-dups (seq-copy (if (ring-p history)
+                                                (ring-elements history)
+                                              history))))
 
 ;; This command has been adopted from https://github.com/oantolin/completing-history/.
 ;;;###autoload
 (defun consult-history (&optional history)
   "Insert string from buffer HISTORY."
   (interactive)
   (let* ((enable-recursive-minibuffers t)
-         (str (consult--read "History: "
-                             (or (consult--history-elements
-                                  (or history (consult--current-history)))
-                                 (user-error "History is empty"))
-                             :history t ;; disable history
-                             :sort nil)))
+         (dup-cands-and-index-var (if history
+                                      (cons history 'nil)
+                                    (consult--current-history)))
+         (dup-cands (car dup-cands-and-index-var))
+         (index-var (cdr dup-cands-and-index-var))
```

Comment:
It would be nice if we could use `dash.el` here to avoid these infelicities.

Reply:
No, I don't want to add a dash dependency. I think the standard library seq, map and cl is rich enough.